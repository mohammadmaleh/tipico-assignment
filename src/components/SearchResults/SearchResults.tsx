import * as React from "react";
import { gitHubRepositoriesState } from "../../types";
import { Pagination, Table, Message } from "semantic-ui-react";
const { HeaderCell, Cell, Body, Row, Header } = Table;
interface ISearchResultsProps {
  gitHubRepositories: gitHubRepositoriesState;
  handleChangePage: (
    event: React.MouseEvent<HTMLAnchorElement>,
    data: any
  ) => void;
  activePage: number;
}
const styles = {
  container: {
    height: "calc(100vh - 70px)",
    display: "flex",
    justifyContent: "center",
    alignContent: "center",
    padding: "30px "
  },
  tableContainer: {
    width: "100%"
  },
  message: {
    width: "100%",
    maxHeight: 100
  }
};
const SearchResults: React.FunctionComponent<ISearchResultsProps> = props => {
  const { gitHubRepositories, handleChangePage, activePage } = props;
  const totalPages =
    Math.floor(gitHubRepositories.total_count / 10) > 100
      ? 100
      : Math.floor(gitHubRepositories.total_count / 10);
  const renderTableRows = () => {
    if (gitHubRepositories && gitHubRepositories.items)
      return gitHubRepositories.items.map(item => (
        <Row key={item.id} data-test="results-table-row">
          <Cell>{item.name}</Cell>
          <Cell>{item.language}</Cell>
          <Cell>{item.owner.login}</Cell>
          <Cell>{item.watchers_count}</Cell>
        </Row>
      ));
    else return null;
  };

  return (
    <div data-test="component-search-results" style={styles.container}>
      {!gitHubRepositories.items && (
        <Message style={styles.message} data-test="welcome-message">
          <Message.Header>Welcome to our github search service!</Message.Header>
          <p>start by writing a repository name to search</p>
        </Message>
      )}
      {gitHubRepositories.items && gitHubRepositories.items.length === 0 && (
        <Message style={styles.message} data-test="no-results-message">
          <Message.Header>Ops!!</Message.Header>
          <p>your search had no results!!</p>
        </Message>
      )}
      {gitHubRepositories.items && gitHubRepositories.items.length > 0 && (
        <div style={styles.tableContainer}>
          <Table celled data-test="results-table">
            <Header>
              <Row>
                <HeaderCell>Repository Name</HeaderCell>
                <HeaderCell>Language</HeaderCell>
                <HeaderCell>Owner</HeaderCell>
                <HeaderCell>star</HeaderCell>
              </Row>
            </Header>
            <Body>{renderTableRows()}</Body>
          </Table>
          <Pagination
            onPageChange={handleChangePage}
            defaultActivePage={activePage}
            totalPages={totalPages}
          />
        </div>
      )}
    </div>
  );
};

export default SearchResults;
