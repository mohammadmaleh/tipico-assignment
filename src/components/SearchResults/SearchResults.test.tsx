import * as React from "react";
import { shallow } from "enzyme";
import SearchResults from "./SearchResults";
import { findByTestAttr } from "../../../test/testUtils";
import { dummyData } from "../../constants";
const defaultProps = {
  handleChangePage: () => {},
  gitHubRepositories: {
    incomplete_results: false,
    items: undefined,
    total_count: 0
  },
  activePage: 1
};

const setup = (props = {}) => {
  const setupProps = { ...defaultProps, ...props };
  return shallow(<SearchResults {...setupProps} />);
};
describe("search form", () => {
  test("it renders without error", () => {
    const wrapper = setup();
    const component = findByTestAttr(wrapper, "component-search-results");
    expect(component.length).toBe(1);
    expect(component).toMatchSnapshot();
  });
  test("render welcome message on init", () => {
    const wrapper = setup({
      gitHubRepositories: {
        incomplete_results: false,
        items: undefined,
        total_count: 0
      }
    });
    const component = findByTestAttr(wrapper, "welcome-message");
    expect(component.length).toBe(1);
    expect(component).toMatchSnapshot();
  });
  test("renders no results found message when receiving an empty array", () => {
    const wrapper = setup({
      gitHubRepositories: {
        incomplete_results: false,
        items: [],
        total_count: 0
      }
    });
    const component = findByTestAttr(wrapper, "no-results-message");
    expect(component.length).toBe(1);
    expect(component).toMatchSnapshot();
  });
  test("renders table when there are results and renders rows  ", () => {
    const wrapper = setup({ gitHubRepositories: dummyData });
    const table = findByTestAttr(wrapper, "results-table");
    expect(table.length).toBe(1);
    expect(table).toMatchSnapshot();
    const row = findByTestAttr(wrapper, "results-table-row");
    expect(row.length).toBe(dummyData.items.length);
    expect(row).toMatchSnapshot();
  });
});
