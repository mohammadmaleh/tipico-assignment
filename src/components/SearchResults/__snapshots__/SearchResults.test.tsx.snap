// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search form it renders without error 1`] = `ShallowWrapper {}`;

exports[`search form render welcome message on init 1`] = `ShallowWrapper {}`;

exports[`search form renders no results found message when receiving an empty array 1`] = `ShallowWrapper {}`;

exports[`search form renders table when there are results  1`] = `ShallowWrapper {}`;

exports[`search form renders table when there are results and renders rows   1`] = `ShallowWrapper {}`;

exports[`search form renders table when there are results and renders rows   2`] = `ShallowWrapper {}`;
